File: android/app/src/main/java/com/golfiq/hud/arhud/ARHUDOverlayView.kt
package com.golfiq.hud.arhud

import android.content.Context
import android.graphics.drawable.GradientDrawable
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import java.util.Locale

class ARHUDOverlayView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : FrameLayout(context, attrs) {

    val calibrateButton: Button = Button(context).apply { text = "Aim → Calibrate" }
    val recenterButton: Button = Button(context).apply { text = "Re-center" }
    val markButton: Button = Button(context).apply { text = "Mark" }
    val fieldRunStartButton: Button = Button(context).apply { text = "Start 9-hole" }
    val fieldRunNextButton: Button = Button(context).apply { text = "Next hole" }
    val fieldRunEndButton: Button = Button(context).apply { text = "End run" }
    private val statusLabel: TextView = TextView(context)
    private val frontLabel: TextView = TextView(context)
    private val centerLabel: TextView = TextView(context)
    private val backLabel: TextView = TextView(context)
    private val modeBadge: TextView = TextView(context)
    private val qaContainer: LinearLayout = LinearLayout(context)
    private val qaFpsLabel: TextView = TextView(context)
    private val qaLatencyLabel: TextView = TextView(context)
    private val qaTrackingLabel: TextView = TextView(context)
    private val qaEtagLabel: TextView = TextView(context)
    private val qaHoleLabel: TextView = TextView(context)
    private val qaRecenterLabel: TextView = TextView(context)
    private val qaPlaysLikeHeader: TextView = TextView(context)
    private val qaPlaysLikeDistanceLabel: TextView = TextView(context)
    private val qaPlaysLikeDeltaHLabel: TextView = TextView(context)
    private val qaPlaysLikeWindLabel: TextView = TextView(context)
    private val qaPlaysLikeKsLabel: TextView = TextView(context)
    private val qaPlaysLikeAlphaHeadLabel: TextView = TextView(context)
    private val qaPlaysLikeAlphaTailLabel: TextView = TextView(context)
    private val qaPlaysLikeEffLabel: TextView = TextView(context)
    private val qaPlaysLikeQualityLabel: TextView = TextView(context)
    private val playsLikeContainer: LinearLayout = LinearLayout(context)
    private val playsLikeHeadline: TextView = TextView(context)
    private val playsLikeDeltaLabel: TextView = TextView(context)
    private val playsLikeChipRow: LinearLayout = LinearLayout(context)
    private val playsLikeSlopeChip: TextView = TextView(context)
    private val playsLikeWindChip: TextView = TextView(context)
    private val playsLikeQualityBadge: TextView = TextView(context)

    enum class Mode {
        GEOSPATIAL,
        COMPASS,
    }

    init {
        setWillNotDraw(false)
        foregroundGravity = Gravity.BOTTOM

        statusLabel.apply {
            text = "Align with pin, then calibrate"
            setTextColor(ContextCompat.getColor(context, android.R.color.white))
            textSize = 12f
            gravity = Gravity.CENTER
        }

        modeBadge.apply {
            text = "Compass"
            setTextColor(ContextCompat.getColor(context, android.R.color.black))
            textSize = 12f
            val density = resources.displayMetrics.density
            val horizontal = (16 * density).toInt()
            val vertical = (6 * density).toInt()
            setPadding(horizontal, vertical, horizontal, vertical)
            background = GradientDrawable().apply {
                cornerRadius = 16 * density
                setColor(ContextCompat.getColor(context, android.R.color.white))
                alpha = (0.9f * 255).toInt()
            }
        }

        val distanceContainer = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.END
        }

        listOf(frontLabel to "F:", centerLabel to "C:", backLabel to "B:").forEach { (label, prefix) ->
            label.apply {
                text = "$prefix --"
                setTextColor(ContextCompat.getColor(context, android.R.color.white))
                textSize = 16f
            }
            distanceContainer.addView(label)
        }

        configurePlaysLike(distanceContainer)

        val controls = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER_HORIZONTAL
            addView(calibrateButton)
            addView(recenterButton)
        }

        val stack = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            val density = resources.displayMetrics.density
            val horizontal = (16 * density).toInt()
            val vertical = (24 * density).toInt()
            setPadding(horizontal, vertical, horizontal, vertical * 2)
            gravity = Gravity.CENTER_HORIZONTAL or Gravity.BOTTOM
            addView(statusLabel)
            addView(controls)
            addView(distanceContainer)
        }

        addView(stack, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT).apply {
            gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
        })
        addView(modeBadge, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT).apply {
            gravity = Gravity.TOP or Gravity.START
            val density = resources.displayMetrics.density
            val margin = (12 * density).toInt()
            setMargins(margin, margin, margin, margin)
        })

        configureQaOverlay()
    }

    fun updateStatus(text: String) {
        post { statusLabel.text = text }
    }

    fun updateDistances(front: String, center: String, back: String) {
        post {
            frontLabel.text = "F: $front"
            centerLabel.text = "C: $center"
            backLabel.text = "B: $back"
        }
    }

    fun updateModeBadge(mode: Mode) {
        val label = when (mode) {
            Mode.GEOSPATIAL -> "Geospatial"
            Mode.COMPASS -> "Compass"
        }
        post { modeBadge.text = label }
    }

    fun setPlaysLikeVisible(visible: Boolean) {
        post { playsLikeContainer.visibility = if (visible) View.VISIBLE else View.GONE }
    }

    fun updatePlaysLike(distanceEff: Double, slope: Double, wind: Double, quality: String) {
        post {
            val total = slope + wind
            playsLikeDeltaLabel.text = String.format(
                Locale.US,
                "Plays-like: %.1f m (Δ %+.1f m)",
                distanceEff,
                total,
            )
            playsLikeSlopeChip.text = String.format(Locale.US, "slope %+.1f m", slope)
            playsLikeWindChip.text = String.format(Locale.US, "wind %+.1f m", wind)
            updateQualityBadge(quality)
        }
    }

    fun updatePlaysLikeQa(
        distanceMeters: Double,
        deltaHMeters: Double,
        windParallel: Double,
        kS: Double,
        alphaHead: Double,
        alphaTail: Double,
        eff: Double,
        quality: String,
    ) {
        post {
            qaPlaysLikeDistanceLabel.text = String.format(Locale.US, "D: %.1f m", distanceMeters)
            qaPlaysLikeDeltaHLabel.text = String.format(Locale.US, "Δh: %+0.1f m", deltaHMeters)
            qaPlaysLikeWindLabel.text = String.format(Locale.US, "W∥: %+0.1f m/s", windParallel)
            qaPlaysLikeKsLabel.text = String.format(Locale.US, "kS: %.2f", kS)
            qaPlaysLikeAlphaHeadLabel.text = String.format(Locale.US, "α_head: %.3f /mph", alphaHead)
            qaPlaysLikeAlphaTailLabel.text = String.format(Locale.US, "α_tail: %.3f /mph", alphaTail)
            qaPlaysLikeEffLabel.text = String.format(Locale.US, "Eff: %.1f m", eff)
            qaPlaysLikeQualityLabel.text = String.format(
                Locale.US,
                "Quality: %s",
                quality.uppercase(Locale.US),
            )
        }
    }

    private fun configurePlaysLike(parent: LinearLayout) {
        val density = resources.displayMetrics.density
        playsLikeContainer.orientation = LinearLayout.VERTICAL
        playsLikeContainer.visibility = View.GONE
        playsLikeContainer.setPadding(0, (8 * density).toInt(), 0, 0)

        playsLikeHeadline.apply {
            text = "Plays-like"
            setTextColor(ContextCompat.getColor(context, android.R.color.white))
            textSize = 14f
        }

        playsLikeQualityBadge.apply {
            text = "--"
            textSize = 11f
            setTextColor(ContextCompat.getColor(context, android.R.color.black))
            val horizontal = (10 * density).toInt()
            val vertical = (4 * density).toInt()
            setPadding(horizontal, vertical, horizontal, vertical)
            background = GradientDrawable().apply {
                cornerRadius = 12 * density
                setColor(ContextCompat.getColor(context, android.R.color.white))
                alpha = (0.85f * 255).toInt()
            }
        }

        playsLikeDeltaLabel.apply {
            text = "Plays-like: --"
            setTextColor(ContextCompat.getColor(context, android.R.color.darker_gray))
            textSize = 12f
        }

        playsLikeChipRow.orientation = LinearLayout.HORIZONTAL
        playsLikeChipRow.setPadding(0, (6 * density).toInt(), 0, 0)

        listOf(playsLikeSlopeChip, playsLikeWindChip).forEach { chip ->
            chip.text = "--"
            chip.textSize = 11f
            chip.setTextColor(ContextCompat.getColor(context, android.R.color.white))
            val horizontal = (10 * density).toInt()
            val vertical = (4 * density).toInt()
            chip.setPadding(horizontal, vertical, horizontal, vertical)
            chip.background = GradientDrawable().apply {
                cornerRadius = 999f
                setColor(ContextCompat.getColor(context, android.R.color.darker_gray))
                alpha = (0.5f * 255).toInt()
            }
        }

        playsLikeChipRow.addView(playsLikeSlopeChip)
        playsLikeChipRow.addView(playsLikeWindChip)

        val headerRow = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.END or Gravity.CENTER_VERTICAL
            addView(playsLikeHeadline)
            addView(playsLikeQualityBadge, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT).apply {
                leftMargin = (8 * density).toInt()
            })
        }

        playsLikeContainer.addView(headerRow)
        playsLikeContainer.addView(playsLikeDeltaLabel)
        playsLikeContainer.addView(playsLikeChipRow)

        parent.addView(playsLikeContainer)
    }

    private fun updateQualityBadge(quality: String) {
        val normalized = quality.lowercase(Locale.US)
        val (label, colorRes) = when (normalized) {
            "good" -> "GOOD" to android.R.color.holo_green_light
            "warn" -> "WARN" to android.R.color.holo_orange_light
            else -> "LOW" to android.R.color.holo_red_light
        }
        playsLikeQualityBadge.text = label
        val density = resources.displayMetrics.density
        val drawable = playsLikeQualityBadge.background as? GradientDrawable
            ?: GradientDrawable().apply { cornerRadius = 12 * density }
        drawable.setColor(ContextCompat.getColor(context, colorRes))
        drawable.alpha = (0.85f * 255).toInt()
        playsLikeQualityBadge.background = drawable
    }

    private fun configureQaOverlay() {
        val density = resources.displayMetrics.density
        qaContainer.apply {
            orientation = LinearLayout.VERTICAL
            val padding = (12 * density).toInt()
            setPadding(padding, padding, padding, padding)
            background = GradientDrawable().apply {
                cornerRadius = 12 * density
                setColor(ContextCompat.getColor(context, android.R.color.black))
                alpha = (0.75f * 255).toInt()
            }
            visibility = View.GONE
        }

        val qaTitle = TextView(context).apply {
            text = "Field QA"
            setTextColor(ContextCompat.getColor(context, android.R.color.white))
            textSize = 14f
        }

        listOf(
            qaTitle,
            qaFpsLabel,
            qaLatencyLabel,
            qaTrackingLabel,
            qaEtagLabel,
            qaHoleLabel,
            qaRecenterLabel,
            qaPlaysLikeHeader,
            qaPlaysLikeDistanceLabel,
            qaPlaysLikeDeltaHLabel,
            qaPlaysLikeWindLabel,
            qaPlaysLikeKsLabel,
            qaPlaysLikeAlphaHeadLabel,
            qaPlaysLikeAlphaTailLabel,
            qaPlaysLikeEffLabel,
            qaPlaysLikeQualityLabel,
        ).forEach { label ->
            label.setTextColor(ContextCompat.getColor(context, android.R.color.white))
            label.textSize = 12f
        }

        qaFpsLabel.text = "FPS: --"
        qaLatencyLabel.text = "Latency: --"
        qaTrackingLabel.text = "Tracking: --"
        qaEtagLabel.text = "ETag age: --"
        qaHoleLabel.text = "Hole: –"
        qaRecenterLabel.text = "Recenter marks: 0"
        qaPlaysLikeHeader.apply {
            text = "Plays-like QA"
            textSize = 13f
            val topPadding = (8 * density).toInt()
            setPadding(0, topPadding, 0, 0)
        }
        qaPlaysLikeDistanceLabel.text = "D: --"
        qaPlaysLikeDeltaHLabel.text = "Δh: --"
        qaPlaysLikeWindLabel.text = "W∥: --"
        qaPlaysLikeKsLabel.text = "kS: --"
        qaPlaysLikeAlphaHeadLabel.text = "α_head: --"
        qaPlaysLikeAlphaTailLabel.text = "α_tail: --"
        qaPlaysLikeEffLabel.text = "Eff: --"
        qaPlaysLikeQualityLabel.text = "Quality: --"

        val markRow = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            addView(markButton)
        }

        val runButtons = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            val spacing = (6 * density).toInt()
            setPadding(0, spacing, 0, 0)
            addView(fieldRunStartButton)
            addView(fieldRunNextButton)
            addView(fieldRunEndButton)
        }

        listOf(markButton, fieldRunStartButton, fieldRunNextButton, fieldRunEndButton).forEach { button ->
            button.textSize = 12f
        }

        fieldRunNextButton.isEnabled = false
        fieldRunEndButton.isEnabled = false
        markButton.isEnabled = false

        qaContainer.addView(qaTitle)
        qaContainer.addView(qaFpsLabel)
        qaContainer.addView(qaLatencyLabel)
        qaContainer.addView(qaTrackingLabel)
        qaContainer.addView(qaEtagLabel)
        qaContainer.addView(qaHoleLabel)
        qaContainer.addView(qaRecenterLabel)
        qaContainer.addView(qaPlaysLikeHeader)
        qaContainer.addView(qaPlaysLikeDistanceLabel)
        qaContainer.addView(qaPlaysLikeDeltaHLabel)
        qaContainer.addView(qaPlaysLikeWindLabel)
        qaContainer.addView(qaPlaysLikeKsLabel)
        qaContainer.addView(qaPlaysLikeAlphaHeadLabel)
        qaContainer.addView(qaPlaysLikeAlphaTailLabel)
        qaContainer.addView(qaPlaysLikeEffLabel)
        qaContainer.addView(qaPlaysLikeQualityLabel)
        qaContainer.addView(markRow)
        qaContainer.addView(runButtons)

        addView(qaContainer, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT).apply {
            gravity = Gravity.TOP or Gravity.END
            val margin = (12 * density).toInt()
            setMargins(margin, margin, margin, margin)
        })
    }

    fun setFieldTestVisible(visible: Boolean) {
        post { qaContainer.visibility = if (visible) View.VISIBLE else View.GONE }
    }

    fun updateFieldTestFps(fps: String) {
        post { qaFpsLabel.text = "FPS: $fps" }
    }

    fun updateFieldTestLatency(label: String) {
        post { qaLatencyLabel.text = "Latency: $label" }
    }

    fun updateFieldTestTracking(label: String) {
        post { qaTrackingLabel.text = "Tracking: $label" }
    }

    fun updateFieldTestEtagAge(days: Int?) {
        val value = when {
            days == null -> "–"
            days <= 0 -> "<1d"
            else -> "${days}d"
        }
        post { qaEtagLabel.text = "ETag age: $value" }
    }

    fun updateFieldRunState(active: Boolean, currentHole: Int?, recenterCount: Int) {
        val holeLabel = if (active && currentHole != null) {
            "Hole: $currentHole/9"
        } else {
            "Hole: –"
        }
        post {
            qaHoleLabel.text = holeLabel
            qaRecenterLabel.text = "Recenter marks: $recenterCount"
            markButton.isEnabled = active
            fieldRunStartButton.isEnabled = !active
            fieldRunNextButton.isEnabled = active && (currentHole ?: 1) < 9
            fieldRunEndButton.isEnabled = active
        }
    }
}

File: ios/App/ARHUD/ARHUDOverlayView.swift
import Foundation
import UIKit

final class ARHUDOverlayView: UIView {
    let calibrateButton: UIButton = UIButton(type: .system)
    let recenterButton: UIButton = UIButton(type: .system)
    let markButton: UIButton = UIButton(type: .system)
    let fieldRunStartButton: UIButton = UIButton(type: .system)
    let fieldRunNextButton: UIButton = UIButton(type: .system)
    let fieldRunEndButton: UIButton = UIButton(type: .system)
    private let statusLabel: UILabel = UILabel()
    private let frontLabel: UILabel = UILabel()
    private let centerLabel: UILabel = UILabel()
    private let backLabel: UILabel = UILabel()
    private let modeBadge: UILabel = UILabel()
    private let qaContainer: UIView = UIView()
    private let qaStack: UIStackView = UIStackView()
    private let qaFpsLabel: UILabel = UILabel()
    private let qaLatencyLabel: UILabel = UILabel()
    private let qaTrackingLabel: UILabel = UILabel()
    private let qaEtagLabel: UILabel = UILabel()
    private let qaHoleLabel: UILabel = UILabel()
    private let qaRecenterLabel: UILabel = UILabel()
    private let qaPlaysLikeHeader: UILabel = UILabel()
    private let qaPlaysLikeDistanceLabel: UILabel = UILabel()
    private let qaPlaysLikeDeltaHLabel: UILabel = UILabel()
    private let qaPlaysLikeWindLabel: UILabel = UILabel()
    private let qaPlaysLikeKsLabel: UILabel = UILabel()
    private let qaPlaysLikeAlphaHeadLabel: UILabel = UILabel()
    private let qaPlaysLikeAlphaTailLabel: UILabel = UILabel()
    private let qaPlaysLikeEffLabel: UILabel = UILabel()
    private let qaPlaysLikeQualityLabel: UILabel = UILabel()
    private let playsLikeContainer: UIStackView = UIStackView()
    private let playsLikeHeadline: UILabel = UILabel()
    private let playsLikeDeltaLabel: UILabel = UILabel()
    private let playsLikeChipStack: UIStackView = UIStackView()
    private let playsLikeSlopeChip: UILabel = UILabel()
    private let playsLikeWindChip: UILabel = UILabel()
    private let playsLikeQualityBadge: UILabel = UILabel()

    enum Mode {
        case geospatial
        case compass
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        configureViewHierarchy()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        configureViewHierarchy()
    }

    private func configureViewHierarchy() {
        backgroundColor = .clear

        calibrateButton.setTitle("Aim → Calibrate", for: .normal)
        calibrateButton.titleLabel?.font = UIFont.preferredFont(forTextStyle: .headline)

        recenterButton.setTitle("Re-center", for: .normal)
        recenterButton.titleLabel?.font = UIFont.preferredFont(forTextStyle: .subheadline)

        markButton.setTitle("Mark", for: .normal)
        markButton.titleLabel?.font = UIFont.preferredFont(forTextStyle: .caption1)

        fieldRunStartButton.setTitle("Start 9-hole", for: .normal)
        fieldRunStartButton.titleLabel?.font = UIFont.preferredFont(forTextStyle: .caption1)

        fieldRunNextButton.setTitle("Next hole", for: .normal)
        fieldRunNextButton.titleLabel?.font = UIFont.preferredFont(forTextStyle: .caption1)
        fieldRunNextButton.isEnabled = false

        fieldRunEndButton.setTitle("End run", for: .normal)
        fieldRunEndButton.titleLabel?.font = UIFont.preferredFont(forTextStyle: .caption1)
        fieldRunEndButton.isEnabled = false

        statusLabel.font = UIFont.preferredFont(forTextStyle: .footnote)
        statusLabel.textColor = .white
        statusLabel.numberOfLines = 2
        statusLabel.textAlignment = .center
        statusLabel.text = "Align device to pin and calibrate"

        modeBadge.font = UIFont.preferredFont(forTextStyle: .caption2)
        modeBadge.textColor = .black
        modeBadge.backgroundColor = UIColor.white.withAlphaComponent(0.85)
        modeBadge.layer.cornerRadius = 8
        modeBadge.layer.masksToBounds = true
        modeBadge.textAlignment = .center
        modeBadge.text = "Compass"
        modeBadge.setContentHuggingPriority(.required, for: .horizontal)
        modeBadge.setContentCompressionResistancePriority(.required, for: .horizontal)

        let distanceStack = UIStackView(arrangedSubviews: [frontLabel, centerLabel, backLabel])
        distanceStack.axis = .vertical
        distanceStack.spacing = 4

        [frontLabel, centerLabel, backLabel].forEach { label in
            label.font = UIFont.monospacedDigitSystemFont(ofSize: 16, weight: .medium)
            label.textColor = .white
            label.textAlignment = .right
            label.text = "--"
        }

        configurePlaysLikeSection(into: distanceStack)

        let controlsStack = UIStackView(arrangedSubviews: [calibrateButton, recenterButton])
        controlsStack.axis = .vertical
        controlsStack.spacing = 8

        let container = UIStackView(arrangedSubviews: [controlsStack, distanceStack])
        container.axis = .vertical
        container.spacing = 16
        container.alignment = .fill

        addSubview(container)
        addSubview(statusLabel)
        addSubview(modeBadge)
        configureFieldTestOverlay()

        container.translatesAutoresizingMaskIntoConstraints = false
        statusLabel.translatesAutoresizingMaskIntoConstraints = false
        modeBadge.translatesAutoresizingMaskIntoConstraints = false

        NSLayoutConstraint.activate([
            container.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor, constant: 16),
            container.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor, constant: -16),
            container.bottomAnchor.constraint(equalTo: safeAreaLayoutGuide.bottomAnchor, constant: -24),

            statusLabel.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor, constant: 16),
            statusLabel.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor, constant: -16),
            statusLabel.bottomAnchor.constraint(equalTo: container.topAnchor, constant: -12),

            modeBadge.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor, constant: 12),
            modeBadge.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor, constant: 16),
            modeBadge.widthAnchor.constraint(greaterThanOrEqualToConstant: 88)
        ])
    }

    func updateStatus(_ text: String) {
        DispatchQueue.main.async {
            self.statusLabel.text = text
        }
    }

    func updateModeBadge(_ mode: Mode) {
        DispatchQueue.main.async {
            switch mode {
            case .geospatial:
                self.modeBadge.text = " Geospatial "
            case .compass:
                self.modeBadge.text = " Compass "
            }
        }
    }

    func updateDistances(front: String, center: String, back: String) {
        DispatchQueue.main.async {
            self.frontLabel.text = "F: \(front)"
            self.centerLabel.text = "C: \(center)"
            self.backLabel.text = "B: \(back)"
        }
    }

    func setPlaysLikeVisible(_ visible: Bool) {
        DispatchQueue.main.async {
            self.playsLikeContainer.isHidden = !visible
        }
    }

    func updatePlaysLike(effective: Double, slope: Double, wind: Double, quality: String) {
        DispatchQueue.main.async {
            let total = slope + wind
            self.playsLikeDeltaLabel.text = String(format: "Plays-like: %.1f m (Δ %+0.1f m)", effective, total)
            self.playsLikeSlopeChip.text = String(format: " slope %+0.1f m ", slope)
            self.playsLikeWindChip.text = String(format: " wind %+0.1f m ", wind)
            self.updateQualityBadge(quality: quality)
        }
    }

    func updatePlaysLikeQA(
        distance: Double,
        deltaH: Double,
        windParallel: Double,
        kS: Double,
        alphaHead: Double,
        alphaTail: Double,
        eff: Double,
        quality: String
    ) {
        DispatchQueue.main.async {
            self.qaPlaysLikeDistanceLabel.text = String(format: "D: %.1f m", distance)
            self.qaPlaysLikeDeltaHLabel.text = String(format: "Δh: %+0.1f m", deltaH)
            self.qaPlaysLikeWindLabel.text = String(format: "W∥: %+0.1f m/s", windParallel)
            self.qaPlaysLikeKsLabel.text = String(format: "kS: %.2f", kS)
            self.qaPlaysLikeAlphaHeadLabel.text = String(format: "α_head: %.3f /mph", alphaHead)
            self.qaPlaysLikeAlphaTailLabel.text = String(format: "α_tail: %.3f /mph", alphaTail)
            self.qaPlaysLikeEffLabel.text = String(format: "Eff: %.1f m", eff)
            self.qaPlaysLikeQualityLabel.text = "Quality: \(quality.uppercased())"
        }
    }

    private func configurePlaysLikeSection(into stack: UIStackView) {
        playsLikeContainer.axis = .vertical
        playsLikeContainer.spacing = 4
        playsLikeContainer.isHidden = true

        let header = UIStackView(arrangedSubviews: [playsLikeHeadline, playsLikeQualityBadge])
        header.axis = .horizontal
        header.spacing = 8
        header.alignment = .center

        playsLikeHeadline.text = "Plays-like"
        playsLikeHeadline.font = UIFont.preferredFont(forTextStyle: .subheadline)
        playsLikeHeadline.textColor = .white

        playsLikeQualityBadge.font = UIFont.preferredFont(forTextStyle: .caption2)
        playsLikeQualityBadge.textAlignment = .center
        playsLikeQualityBadge.text = "--"
        playsLikeQualityBadge.textColor = .black
        playsLikeQualityBadge.backgroundColor = UIColor.white.withAlphaComponent(0.85)
        playsLikeQualityBadge.layer.cornerRadius = 8
        playsLikeQualityBadge.layer.masksToBounds = true
        playsLikeQualityBadge.setContentHuggingPriority(.required, for: .horizontal)

        playsLikeDeltaLabel.text = "Plays-like: --"
        playsLikeDeltaLabel.font = UIFont.preferredFont(forTextStyle: .caption1)
        playsLikeDeltaLabel.textColor = UIColor(white: 0.8, alpha: 1)
        playsLikeDeltaLabel.textAlignment = .right

        playsLikeChipStack.axis = .horizontal
        playsLikeChipStack.spacing = 6

        [playsLikeSlopeChip, playsLikeWindChip].forEach { label in
            label.font = UIFont.preferredFont(forTextStyle: .caption2)
            label.textColor = .white
            label.textAlignment = .center
            label.text = " -- "
            label.backgroundColor = UIColor.white.withAlphaComponent(0.25)
            label.layer.cornerRadius = 10
            label.layer.masksToBounds = true
            label.setContentHuggingPriority(.required, for: .horizontal)
            label.setContentCompressionResistancePriority(.required, for: .horizontal)
            label.translatesAutoresizingMaskIntoConstraints = false
            label.heightAnchor.constraint(greaterThanOrEqualToConstant: 20).isActive = true
        }

        playsLikeChipStack.addArrangedSubview(playsLikeSlopeChip)
        playsLikeChipStack.addArrangedSubview(playsLikeWindChip)

        playsLikeContainer.addArrangedSubview(header)
        playsLikeContainer.addArrangedSubview(playsLikeDeltaLabel)
        playsLikeContainer.addArrangedSubview(playsLikeChipStack)

        stack.addArrangedSubview(playsLikeContainer)
    }

    private func updateQualityBadge(quality: String) {
        let normalized = quality.lowercased()
        switch normalized {
        case "good":
            playsLikeQualityBadge.text = "GOOD"
            playsLikeQualityBadge.backgroundColor = UIColor.systemGreen.withAlphaComponent(0.85)
        case "warn":
            playsLikeQualityBadge.text = "WARN"
            playsLikeQualityBadge.backgroundColor = UIColor.systemOrange.withAlphaComponent(0.85)
        default:
            playsLikeQualityBadge.text = "LOW"
            playsLikeQualityBadge.backgroundColor = UIColor.systemRed.withAlphaComponent(0.85)
        }
    }

    private func configureFieldTestOverlay() {
        qaContainer.translatesAutoresizingMaskIntoConstraints = false
        qaContainer.backgroundColor = UIColor.black.withAlphaComponent(0.7)
        qaContainer.layer.cornerRadius = 12
        qaContainer.layer.masksToBounds = true
        qaContainer.isHidden = true

        qaStack.axis = .vertical
        qaStack.spacing = 4
        qaStack.alignment = .leading
        qaStack.translatesAutoresizingMaskIntoConstraints = false

        let titleLabel = UILabel()
        titleLabel.text = "Field QA"
        titleLabel.font = UIFont.preferredFont(forTextStyle: .caption1)
        titleLabel.textColor = .white

        [qaFpsLabel, qaLatencyLabel, qaTrackingLabel, qaEtagLabel, qaHoleLabel, qaRecenterLabel].forEach { label in
            label.font = UIFont.preferredFont(forTextStyle: .caption2)
            label.textColor = .white
        }
        [
            qaPlaysLikeHeader,
            qaPlaysLikeDistanceLabel,
            qaPlaysLikeDeltaHLabel,
            qaPlaysLikeWindLabel,
            qaPlaysLikeKsLabel,
            qaPlaysLikeAlphaHeadLabel,
            qaPlaysLikeAlphaTailLabel,
            qaPlaysLikeEffLabel,
            qaPlaysLikeQualityLabel
        ].forEach { label in
            label.font = UIFont.preferredFont(forTextStyle: .caption2)
            label.textColor = .white
        }
        qaFpsLabel.text = "FPS: --"
        qaLatencyLabel.text = "Latency: --"
        qaTrackingLabel.text = "Tracking: --"
        qaEtagLabel.text = "ETag age: --"
        qaHoleLabel.text = "Hole: –"
        qaRecenterLabel.text = "Recenter marks: 0"
        qaPlaysLikeHeader.text = "Plays-like QA"
        qaPlaysLikeHeader.font = UIFont.preferredFont(forTextStyle: .caption1)
        qaPlaysLikeDistanceLabel.text = "D: --"
        qaPlaysLikeDeltaHLabel.text = "Δh: --"
        qaPlaysLikeWindLabel.text = "W∥: --"
        qaPlaysLikeKsLabel.text = "kS: --"
        qaPlaysLikeAlphaHeadLabel.text = "α_head: --"
        qaPlaysLikeAlphaTailLabel.text = "α_tail: --"
        qaPlaysLikeEffLabel.text = "Eff: --"
        qaPlaysLikeQualityLabel.text = "Quality: --"

        let markStack = UIStackView(arrangedSubviews: [markButton])
        markStack.axis = .vertical
        markStack.spacing = 4

        let runButtons = UIStackView(arrangedSubviews: [fieldRunStartButton, fieldRunNextButton, fieldRunEndButton])
        runButtons.axis = .vertical
        runButtons.spacing = 4

        qaStack.addArrangedSubview(titleLabel)
        qaStack.addArrangedSubview(qaFpsLabel)
        qaStack.addArrangedSubview(qaLatencyLabel)
        qaStack.addArrangedSubview(qaTrackingLabel)
        qaStack.addArrangedSubview(qaEtagLabel)
        qaStack.addArrangedSubview(qaHoleLabel)
        qaStack.addArrangedSubview(qaRecenterLabel)
        qaStack.addArrangedSubview(qaPlaysLikeHeader)
        qaStack.addArrangedSubview(qaPlaysLikeDistanceLabel)
        qaStack.addArrangedSubview(qaPlaysLikeDeltaHLabel)
        qaStack.addArrangedSubview(qaPlaysLikeWindLabel)
        qaStack.addArrangedSubview(qaPlaysLikeKsLabel)
        qaStack.addArrangedSubview(qaPlaysLikeAlphaHeadLabel)
        qaStack.addArrangedSubview(qaPlaysLikeAlphaTailLabel)
        qaStack.addArrangedSubview(qaPlaysLikeEffLabel)
        qaStack.addArrangedSubview(qaPlaysLikeQualityLabel)
        qaStack.addArrangedSubview(markStack)
        qaStack.addArrangedSubview(runButtons)

        qaContainer.addSubview(qaStack)
        addSubview(qaContainer)

        NSLayoutConstraint.activate([
            qaContainer.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor, constant: 12),
            qaContainer.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor, constant: -16),

            qaStack.leadingAnchor.constraint(equalTo: qaContainer.leadingAnchor, constant: 12),
            qaStack.trailingAnchor.constraint(equalTo: qaContainer.trailingAnchor, constant: -12),
            qaStack.topAnchor.constraint(equalTo: qaContainer.topAnchor, constant: 12),
            qaStack.bottomAnchor.constraint(equalTo: qaContainer.bottomAnchor, constant: -12)
        ])

        markButton.isEnabled = false
    }

    func setFieldTestVisible(_ visible: Bool) {
        DispatchQueue.main.async {
            self.qaContainer.isHidden = !visible
        }
    }

    func updateFieldTestFps(_ fps: String) {
        DispatchQueue.main.async {
            self.qaFpsLabel.text = "FPS: \(fps)"
        }
    }

    func updateFieldTestLatency(_ label: String) {
        DispatchQueue.main.async {
            self.qaLatencyLabel.text = "Latency: \(label)"
        }
    }

    func updateFieldTestTracking(_ label: String) {
        DispatchQueue.main.async {
            self.qaTrackingLabel.text = "Tracking: \(label)"
        }
    }

    func updateFieldTestEtagAge(_ days: Int?) {
        let text: String
        if let days {
            if days <= 0 {
                text = "<1d"
            } else {
                text = "\(days)d"
            }
        } else {
            text = "–"
        }
        DispatchQueue.main.async {
            self.qaEtagLabel.text = "ETag age: \(text)"
        }
    }

    func updateFieldRunState(active: Bool, currentHole: Int?, recenterCount: Int) {
        let holeText: String
        if active, let currentHole {
            holeText = "Hole: \(currentHole)/9"
        } else {
            holeText = "Hole: –"
        }
        DispatchQueue.main.async {
            self.qaHoleLabel.text = holeText
            self.qaRecenterLabel.text = "Recenter marks: \(recenterCount)"
            self.markButton.isEnabled = active
            self.fieldRunStartButton.isEnabled = !active
            self.fieldRunNextButton.isEnabled = active && (currentHole ?? 1) < 9
            self.fieldRunEndButton.isEnabled = active
        }
    }
}
